idx/memory/tag_query.go

```diff
@@ -171,247 +252,751 @@ func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+// are then pushed into the given tag channel
+func (q *TagQuery) filterTagsFromChan(idCh chan idx.MetricID, tagCh chan string, stopCh chan struct{}, matchName bool) {
+	// used to prevent that this worker thread will push the same result into
+	// the chan twice
+	resultsCache := make(map[string]struct{})
+
+IDS:
+	for id := range idCh {
+		var def *idx.Archive
+		var ok bool
+
+		if def, ok = q.byId[id.String()]; !ok {
+			// should never happen because every ID in the tag index
+			// must be present in the byId lookup table
+			corruptIndex.Inc()
+			log.Error(3, "memory-idx: ID %q is in tag index but not in the byId lookup table", id.String())
+			continue
+		}
+
+		// generate a set of all tags of the current metric that satisfy the
+		// tag filter condition
+		metricTags := make(map[string]struct{}, 0)
+		for _, tag := range def.Tags {
+			equal := strings.Index(tag, "=")
+			if equal < 0 {
+				corruptIndex.Inc()
+				log.Error(3, "memory-idx: tag is in index, but does not contain '=' sign: %s", tag)
+				continue
+			}
+
+			key := tag[:equal]
+			// this tag has already been pushed into tagCh, so we can stop evaluating
+			if _, ok := resultsCache[key]; ok {
+				continue
+			}
+
+			if q.filterTagBy == PREFIX_TAG {
+				if !strings.HasPrefix(key, q.tagPrefix) {
+					continue
+				}
+			} else if q.filterTagBy == MATCH_TAG {
+				if _, ok := q.tagMatch.missCache.Load(key); ok || !q.tagMatch.value.MatchString(tag) {
+					if !ok {
+						q.tagMatch.missCache.Store(key, struct{}{})
+					}
+					continue
+				}
+			}
+			metricTags[key] = struct{}{}
+		}
+
+		if matchName {
+			if _, ok := resultsCache["name"]; !ok {
+				metricTags["name"] = struct{}{}
+			}
+		}
+
+		// if some tags satisfy the current tag filter condition then we run
+		// the metric through all tag expression tests in order to decide
+		// whether those tags should be part of the final result set
+		if len(metricTags) > 0 {
+			if q.testByAllExpressions(id, def, matchName) {
+				for key := range metricTags {
+					select {
+					case tagCh <- key:
+					case <-stopCh:
+						// if execution of query has stopped because the max tag
+						// count has been reached then tagCh <- might block
+						// because that channel will not be consumed anymore. in
+						// that case the stop channel will have been closed so
+						// we so we exit here
+						break IDS
+					}
+					resultsCache[key] = struct{}{}
+				}
+			} else {
+				// check if we need to stop
+				select {
+				case <-stopCh:
+					break IDS
+				default:
+				}
+			}
+		}
+	}
+
+	q.wg.Done()
+}
+
+// determines whether the given tag prefix/tag match will match the special
+// tag "name". if it does, then we can omit some filtering because we know
+// that every metric has a name
+func (q *TagQuery) tagFilterMatchesName() bool {
+	matchName := false
+
+	if q.filterTagBy == PREFIX_TAG || q.startWith == PREFIX_TAG {
+		if strings.HasPrefix("name", q.tagPrefix) {
```

Comment:
is it safe to just do a prefix match? maybe somebody is querying for tag namebar or something? ditto for branch below

Reply:
you're right, actually i think it should be like `strings.HasPrefix("name=", q.tagPrefix)` because `=` is not allowed in the tag

Reply:
that should be fine. 
if somebody queries for `namebar` then this should not match, and it won't: https://play.golang.org/p/39M62UOn8Z
if somebody queries for `na` then this should match, and it will:
https://play.golang.org/p/VdZ3cZpkAE

Reply:
basically the idea of this `tagFilterMatchesName()` is just to do the same evaluation on the "virtual" tag `name` that would be done on every tag of every metric that needs to be evaluated. because when we know that the special tag `name` matches the queried pattern, and we know that every single metric has a name, then we know that every metric will pass the test. so we can skip a lot of tests.

Reply:
it's kind of an ugly hack that is supposed to short cut certain cases by taking advantage of the fact that we know that every metric has a name